### DISCUSSION TITLE
Getting a refresh token

### DISCUSSION THREAD

**Discussion by LukePetko** [iTwin/itwinjs-core#4569] [Q&A] [answered]
Hi,

I've got some trouble getting the refresh token from the API. The documentation says: `Refresh token is requested with a special offlineaccess_ scope, and such a request will always present a permissions consent page for end user`. 

I want to ask if I have to add this scope to my scopes used during the sign-in process, or if I should make a new API call.
I'd also like to ask if I have to configure anything on the iTwin Platform.

Thanks

**Comment by calebmshafer** [comment, marked as answer]
@LukePetko do you have the app registered as a Desktop/Mobile application?  Those instructions are a bit old and you do not need to explicitly add that scope if you're using the Desktop/Mobile app registration that gives you a `native-xxxx` client id.

**Reply by LukePetko** [reply]
My app is actually registered as a web app

**Reply by LukePetko** [reply]
Do you think that there is any way to get it also for a SPA?

**Reply by calebmshafer** [reply]
@LukePetko we do not allow the offline scope for SPA since that returns a refresh token in all cases in addition to the user token. Are you building a web application? If so, what workflow are you trying to use that you'd need a refresh token?

**Reply by LukePetko** [reply]
@calebmshafer so SPA always return the refresh token? If so, could you please show me where? I'm using the https://www.npmjs.com/package/@itwin/browser-authorization/v/0.5.1 lib

Thanks

**Reply by evelynpreslar-bentley** [reply]
SPA apps should never receive refresh tokens in any case.

Refresh tokens are long-lived and directly exchangeable for access tokens for the duration, which is why they commonly require client secrets to obtain.
Because SPAs are public-facing, secrets cannot be securely stored, which undermines the security of any refresh tokens issued.
Our SPA client registrations disallow the offline scope to prevent this situation from occurring in the first place.

Instead, SPA apps are encouraged to use [Authorization Code Flow](https://auth0.com/docs/get-started/authentication-and-authorization-flow/authorization-code-flow) to retrieve auth tokens.
This can be done in `@itwin/browser-authorization` by specifying a `responseType` of `code`.
Authorization codes are exchangeable for tokens _once_, not indefinitely, and then only when accompanied by other verifying data, making them far more secure than refresh tokens for public clients like SPAs.

To see the code which performs the code exchange, please refer to the [`oidc-client-js`](https://github.com/IdentityModel/oidc-client-js) library, which is used by `@itwin/browser-authorization` to request tokens.
Code exchange happens during callback handling starting [here](https://github.com/IdentityModel/oidc-client-js/blob/edec8f59897bdeedcb0b4167586d49626203c2c1/src/UserManager.js#L401) and ending [here](https://github.com/IdentityModel/oidc-client-js/blob/edec8f59897bdeedcb0b4167586d49626203c2c1/src/ResponseValidator.js#L228-L230)

**Reply by LukePetko** [reply]
Thanks for the explanation, it cleared out a lot of things to me. Could you please also direct me to some way to get the new `access_token` every time the old one is about to expire? I'm still a bit new to this topic so I do not know any best practices about this.

Thanks

**Reply by evelynpreslar-bentley** [reply]
If using `browser-authorization`, the token is refreshed automatically by the client, and old tokens are discarded when they expire.

Whenever a token is needed for a new request, you can obtain one via [`getAccessToken()`](https://github.com/iTwin/auth-clients/blob/66a5ade18466c439f5e9d3f8c5d687306950104e/packages/browser/src/Client.ts#L299).
There is no need to separately cache the token. The client handles this already, so calling the function every time you need a token ensures that you always received one that's up to date.

This function will return a rejected promise if the user isn't signed in (which can happen simply due to time).
To detect when a user needs to sign in again, you can check the [`isAuthorized`](https://github.com/iTwin/auth-clients/blob/66a5ade18466c439f5e9d3f8c5d687306950104e/packages/browser/src/Client.ts#L90-L92) property before calling the desired signin function.
